File: app/src/main/kotlin/com/gtaclone/ui/VirtualJoystick.kt
package com.gtaclone.ui

import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import kotlin.math.hypot

class VirtualJoystick {

    var centerX = 200f; var centerY = 600f
    private var outerR = 120f; private var innerR = 46f

    var dirX = 0f; var dirY = 0f
    private var thumbX = centerX; private var thumbY = centerY

    private val fillPaint = Paint().apply { color = Color.parseColor("#44FFFFFF"); isAntiAlias = true }
    private val ringPaint = Paint().apply {
        color = Color.parseColor("#99FFFFFF"); style = Paint.Style.STROKE; strokeWidth = 3f; isAntiAlias = true
    }
    private val thumbPaint = Paint().apply { color = Color.parseColor("#CCFFFFFF"); isAntiAlias = true }
    private val arrowPaint = Paint().apply {
        color = Color.parseColor("#AAFFFFFF"); style = Paint.Style.STROKE; strokeWidth = 3f; isAntiAlias = true
    }

    fun init(cx: Float, cy: Float, radius: Float) {
        centerX = cx; centerY = cy; outerR = radius; innerR = radius * 0.38f
        thumbX = cx; thumbY = cy
    }

    fun contains(x: Float, y: Float) = hypot(x - centerX, y - centerY) <= outerR

    fun onTouchDown(x: Float, y: Float) = move(x, y)
    fun onTouchMove(x: Float, y: Float) = move(x, y)
    fun onTouchUp() { thumbX = centerX; thumbY = centerY; dirX = 0f; dirY = 0f }

    private fun move(x: Float, y: Float) {
        val dx = x - centerX; val dy = y - centerY
        val dist = hypot(dx, dy); val max = outerR - innerR
        if (dist > max) { thumbX = centerX + dx / dist * max; thumbY = centerY + dy / dist * max }
        else { thumbX = x; thumbY = y }
        dirX = (thumbX - centerX) / max; dirY = (thumbY - centerY) / max
    }

    fun draw(canvas: Canvas) {
        canvas.drawCircle(centerX, centerY, outerR, fillPaint)
        canvas.drawCircle(centerX, centerY, outerR, ringPaint)
        arrow(canvas, centerX, centerY - outerR * 0.62f, 0f)
        arrow(canvas, centerX, centerY + outerR * 0.62f, 180f)
        arrow(canvas, centerX - outerR * 0.62f, centerY, -90f)
        arrow(canvas, centerX + outerR * 0.62f, centerY, 90f)
        canvas.drawCircle(thumbX, thumbY, innerR, thumbPaint)
    }

    private fun arrow(c: Canvas, x: Float, y: Float, a: Float) {
        c.save(); c.translate(x, y); c.rotate(a)
        c.drawLine(0f, -9f, 0f, 9f, arrowPaint)
        c.drawLine(-7f, 2f, 0f, -7f, arrowPaint)
        c.drawLine(7f, 2f, 0f, -7f, arrowPaint)
        c.restore()
    }
}

File: GTAClone-iOS/GTAClone/ui/VirtualJoystick.swift
import UIKit

class VirtualJoystick {

    var centerX: CGFloat = 200; var centerY: CGFloat = 600
    private var outerR: CGFloat = 120; private var innerR: CGFloat = 46

    private(set) var dirX: Float = 0
    private(set) var dirY: Float = 0
    private var thumbX: CGFloat; private var thumbY: CGFloat

    init() { thumbX = centerX; thumbY = centerY }

    func configure(cx: CGFloat, cy: CGFloat, radius: CGFloat) {
        centerX = cx; centerY = cy; outerR = radius; innerR = radius * 0.38
        thumbX = cx; thumbY = cy
    }

    func contains(_ p: CGPoint) -> Bool { hypot(p.x - centerX, p.y - centerY) <= outerR }

    func onTouchDown(_ p: CGPoint) { move(to: p) }
    func onTouchMove(_ p: CGPoint) { move(to: p) }
    func onTouchUp() { thumbX = centerX; thumbY = centerY; dirX = 0; dirY = 0 }

    private func move(to p: CGPoint) {
        let dx = p.x - centerX; let dy = p.y - centerY
        let dist = hypot(dx, dy); let max = outerR - innerR
        if dist > max {
            thumbX = centerX + (dx/dist)*max; thumbY = centerY + (dy/dist)*max
        } else { thumbX = p.x; thumbY = p.y }
        dirX = Float((thumbX - centerX) / max)
        dirY = Float((thumbY - centerY) / max)
    }

    func draw(in ctx: CGContext) {
        ctx.setFillColor(UIColor(white: 1, alpha: 0.25).cgColor)
        ctx.fillEllipse(in: CGRect(x: centerX-outerR, y: centerY-outerR, width: outerR*2, height: outerR*2))

        ctx.setStrokeColor(UIColor(white: 1, alpha: 0.6).cgColor)
        ctx.setLineWidth(3)
        ctx.strokeEllipse(in: CGRect(x: centerX-outerR, y: centerY-outerR, width: outerR*2, height: outerR*2))

        // Arrows
        let ar = outerR * 0.62
        drawArrow(ctx, x: centerX, y: centerY - ar, angle: 0)
        drawArrow(ctx, x: centerX, y: centerY + ar, angle: .pi)
        drawArrow(ctx, x: centerX - ar, y: centerY, angle: -.pi/2)
        drawArrow(ctx, x: centerX + ar, y: centerY, angle: .pi/2)

        ctx.setFillColor(UIColor(white: 1, alpha: 0.8).cgColor)
        ctx.fillEllipse(in: CGRect(x: thumbX-innerR, y: thumbY-innerR, width: innerR*2, height: innerR*2))
    }

    private func drawArrow(_ ctx: CGContext, x: CGFloat, y: CGFloat, angle: CGFloat) {
        ctx.saveGState()
        ctx.translateBy(x: x, y: y); ctx.rotate(by: angle)
        ctx.setStrokeColor(UIColor(white: 1, alpha: 0.65).cgColor); ctx.setLineWidth(3)
        ctx.move(to: .init(x: 0, y: -9)); ctx.addLine(to: .init(x: 0, y: 9)); ctx.strokePath()
        ctx.move(to: .init(x: -7, y: 2)); ctx.addLine(to: .init(x: 0, y: -7)); ctx.addLine(to: .init(x: 7, y: 2)); ctx.strokePath()
        ctx.restoreGState()
    }
}
